[Popover] prop.marginThreshold can now be array for different horizontal and vertical margins

daniel-rabe opened the pull request:
I already did a pull request some time ago with that feature. We still require this in our applications, is it possible to merge it now?

https://github.com/mui/material-ui/pull/19101

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

mnajdova commented:
I am not sure anything has changed since the last PR, we still plan to use the `Popper` long term. Have you tried using it to achieve the same behavior? Are you able to achieve the same behavior that you want on the `Popover` by using a custom style/override?

daniel-rabe commented:
to make sure the popover is not positioned to height (behind notch/statusbar) we could add marginTop but the position/height check of the component does not work correct then, which results in a popover with bottom part out of screen when the popover has too much content

@mnajdova

mnajdova commented:
If we want to push this further, we will need to:

- add tests
- ensure that the same API would be possible with Popper (and Floating UI), so that when we move to them we don't lack some features

ZeeshanTamboli commented:
@daniel-rabe Any updates?

daniel-rabe commented:
i added some tests

daniel-rabe commented:
when mui switches to popper js the values can still be used

https://popper.js.org/docs/v2/utils/detect-overflow/#padding

ZeeshanTamboli reviewed (changes requested):
Please merge latest master branch, and also fix the tests. The CI is failing.

ZeeshanTamboli reviewed (changes requested):
There are a few things to consider:
- The UI breaks when only a number is provided to marginThreshold. 
  - Before: https://codesandbox.io/s/hungry-hooks-kywcsx?file=/demo.tsx. 
  - After: https://codesandbox.io/s/create-react-app-with-typescript-forked-98ox5m?file=/src/App.tsx
 - What if we want to have only vertical threshold or only horizontal threshold? In this case, I think array with two values won't work. Would an object (top, bottom, left, right) be a better API?
 - We need to to also add test cases to confirm whether the correct styles are getting applied to the Popover in `Popover.test.js`.

daniel-rabe commented:
yes i think an object like padding-option of PopperJS/FloatingUI is better. i will change it later today

daniel-rabe commented:
i changed it to object @ZeeshanTamboli

ZeeshanTamboli reviewed (commented):
> i changed it to object @ZeeshanTamboli

Points 1 & 3 still needs to be considered from https://github.com/mui/material-ui/pull/31994#pullrequestreview-1206467031

daniel-rabe commented:
> > i changed it to object @ZeeshanTamboli
> 
> Points 1 & 3 still needs to be considered from [#31994 (review)](https://github.com/mui/material-ui/pull/31994#pullrequestreview-1206467031)

i updated the tests

ZeeshanTamboli commented:
> > > i changed it to object @ZeeshanTamboli
> > 
> > 
> > Points 1 & 3 still needs to be considered from [#31994 (review)](https://github.com/mui/material-ui/pull/31994#pullrequestreview-1206467031)
> 
> i updated the tests

@daniel-rabe Still need to resolve point 1 mentioned in https://github.com/mui/material-ui/pull/31994#pullrequestreview-1206467031.

ZeeshanTamboli commented:
Closing due to inactivity. And to note, we are going to likely migrate to Floating UI in the future.
